lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,342 @@
+        interfaceKeys.push_back(key);
+      }
+      // Create the interfaces.  These will be populated later.
+      if (interfaces.count(defName.getValue()) == 0) {
+        auto interfaceOp =
+            builder.create<sv::InterfaceOp>(circuitOp->getLoc(), defName);
+        ImplicitLocOpBuilder blockBuilder(interfaceOp->getLoc(),
+                                          interfaceOp.getRegion());
+        blockBuilder.createBlock(&interfaceOp.getRegion());
+        blockBuilder.create<sv::TypeDeclTerminatorOp>();
+        interfaces[defName.getValue()] = interfaceOp;
+      }
+      continue;
+    }
+
+    unprocessedAnnos.push_back(anno.getDict());
+  }
+
+  circuitOp->setAttr("annotations", builder.getArrayAttr(unprocessedAnnos));
+
+  // Walk through the circuit to collect additional information.  If this fails,
+  // signal pass failure.
+  for (auto &op : circuitOp.getBody()->getOperations()) {
+    bool failed =
+        TypeSwitch<Operation *, bool>(&op).Case<FModuleOp, FExtModuleOp>(
+            [&](auto op) {
+              GrandCentralVisitor visitor(&getContext(), &interfaceMap);
+              visitor.visitDecl(op);
+              return visitor.hasFailed();
+            });
+    if (failed)
+      return signalPassFailure();
+  }
+
+  // Populate interfaces, iterating in reverse order as statements are
+  // inserted at the front of the Interface region.
+  auto loc = UnknownLoc::get(context);
+  for (auto i = interfaceKeys.size(); i != 0; --i) {
+    auto a = interfaceKeys[i - 1];
+    auto defName = a.first;
+    auto name = a.second;
+
+    auto info = &interfaceMap[{defName, name}];
+    if (!info->found && info->tpe != Component::Type::Bundle) {
+      circuitOp.emitOpError()
+          << "contained a Grand Central Interface with defName '" << defName
+          << "' that had a component called '" << name
+          << "' which did not have a scattered companion annotation.  This may "
+             "indicate that the companion was improperly deleted.";
+      continue;
+    }
+
+    auto description = info->description;
+
+    auto interfaceOp = interfaces[defName];
+    builder.setInsertionPoint(&interfaceOp.body().front(),
+                              interfaceOp.body().front().begin());
+
+    if (!description.empty())
+      builder.create<sv::VerbatimOp>(loc, ("\n// " + description).str());
+
+    switch (info->tpe) {
+    case Component::Type::Bundle:
+      // TODO: Change this to actually use an interface type.  This currently
+      // does not work because: (1) interfaces don't have a defined way to get
+      // their bit width and (2) interfaces have a symbol table that is used to
+      // verify internal ops, but this requires looking arbitrarily far upwards
+      // to find other symbols.
+      interfaceOp.emitWarning()
+          << "the circuit contains a nested Grand Central interface that will "
+             "be emitted unsafely as verbatim SystemVerilog";
```

Comment:
Why is this warning helpful to users?

I think it would be better to just put this as a `// TODO:` comment.

Reply:
It's totally not. :joy: Will remove this.